debian/coverdienst.1
.TH "COVERDIENST" "1" "" "Manual" ""
.SH NAME
.PP
coverdienst \- Cover images as SeeAlso service
.SH SYNOPSIS
.PP
The application is automatically started as service, listening on port
6027.
.IP
.nf
\f[C]
sudo\ service\ coverdienst\ {status|start|stop|restart}
\f[]
.fi
.SH DESCRIPTION
.PP
Coverdienst provides a SeeAlso web service for cover images.
.IP \[bu] 2
\f[C]format=img\f[] returns a cover image or a transparent 1x1 GIF given
a document identifier
.IP \[bu] 2
\f[C]format=seealso\f[] returns information about a cover image in Open
Search Suggestions / SeeAlso JSON format
.RS 2
.PP
[ id, [ type ], [ size ], [ URL ] ]
.RE
.PP
where \f[C]id\f[] is the normalized given document identifier,
\f[C]type\f[] is \f[C]image/jpeg\f[] or \f[C]image/gif\f[],
\f[C]size\f[] is the image size (e.g.
\f[C]300x400\f[]), and URL is the query URL with \f[C]format=img\f[].
.PP
SeeAlso JSON format can be queried by many means.
A sample JavaScript client is included and served at
\f[C]/coverdienst.json\f[].
.SH INSTALLATION
.PP
The application is packaged as Debian package.
No binaries are included, so the package should work on all
architectures.
It is tested with Ubuntu 14.04 LTS.
.PP
Files are installed at the following locations:
.IP \[bu] 2
\f[C]/srv/coverdienst/\f[] \- application
.IP \[bu] 2
\f[C]/var/log/coverdienst/\f[] \- log files
.IP \[bu] 2
\f[C]/etc/default/coverdienst\f[] \- server configuration
.IP \[bu] 2
\f[C]/etc/coverdienst/\f[] \- application configuration
.PP
Cover images must manually be copied to \f[C]/srv/coverdienst/data/\f[]
and chown\[aq]ed to user \f[C]coverdienst\f[].
.SH CONFIGURATION
.PP
See \f[C]/etc/default/coverdienst\f[] for basic configuration.
Settings are not modified by updates.
Restart is needed after changes.
The following keys are required:
.IP \[bu] 2
\f[C]PORT\f[] \- port number (required, 6027 by default)
.IP \[bu] 2
\f[C]WORKERS\f[] \- number of parallel connections (required, 5 by
default).
If put behind a HTTP proxy, this number is not affected by slow cient
connections but only by the time of processing each request.
.PP
Additional configuration is placed in the JSON file
\f[C]/etc/coverdienst/coverdienst.json\f[].
The following keys are recognized, among others:
.IP \[bu] 2
\f[C]proxy\f[] \- a space\-or\-comma\-separated list of trusted IPs or
IP\-ranges (e.g.
\f[C]192.168.0.0/16\f[]) to take from the \f[C]X\-Forwarded\-For\f[]
header.
The special value \f[C]*\f[] can be used to trust all IPs.
.PP
Image files are located in \f[C]/srv/coverdienst/data/\f[].
The directory \f[C]/srv/coverdienst/bin\f[] contains some maintainance
scripts.
The script \f[C]isbn2file\f[] can be used to map a given ISBN or ISBN
file to the corresponding cover image filename.
.SH SEE ALSO
.PP
Changelog is located in file
\f[C]debian/changelog\f[] (debian/changelog) in the source code
repository.
.PP
Source code and issue tracker at <https://github.com/gbv/coverdienst>.
See file \f[C]CONTRIBUTING.md\f[] (CONTRIBUTING.md) for source code
organization.


debian/changelog
coverdienst (0.15) UNRELEASED; urgency=medium

  * Fixed getting cover from PICA field 009Q

 -- Jakob Voß <jakob.voss@gbv.de>  Wed, 07 Dec 2016 14:12:25 +0100

coverdienst (0.14) unstable; urgency=medium

  * Added coverdienst.json client 
  * Change config file from YAML to JSON
  * Rotate logs

 -- Jakob Voß <jakob.voss@gbv.de>  Mon, 11 May 2015 11:13:23 +0200

coverdienst (0.13) unstable; urgency=medium

  * Change config file from JSON to YAML
  * Support configuration of proxy

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 26 Feb 2015 16:32:47 +0100

coverdienst (0.12) unstable; urgency=medium

  * Move config file to /etc/coverdienst 

    Future updates won't modify the config files

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 05 Feb 2015 16:32:47 +0100

coverdienst (0.11) unstable; urgency=medium

  * Fix getting cover via PICA record

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 05 Feb 2015 15:46:51 +0100

coverdienst (0.10) unstable; urgency=low

  * Debian package

 -- Jakob Voß <jakob.voss@gbv.de>  Mon, 19 Jan 2015 14:47:27 +0100


htdocs/coverdienst.js
/**
 * Coverdienst-Einbindung ohne weitere JavaScript-Abhängigkeiten.
 * Benoetigt mindestens IE 9.0 oder anderen Browser.
 */
function escapeHTML(s) {
    return s.replace(/&/g,"&amp;").replace(/"/g,"&quot;")
            .replace(/</g,"&lt;").replace(/>/g,"&gt;");
}

function displayCover(element, size, url) {
    var wantWidth = 1 * element.getAttribute("width");
    var wantHeight = 1 * element.getAttribute("height");
    var attr = "";
    var dim = size.match(/^(\d+)x(\d+)$/);
    if (dim) {
        var w = dim[1], h = dim[2];
        if (w && h) {
            var width = w, height = h;
            if (wantWidth && !wantHeight) {
                width = wantWidth;
                height = h * (wantWidth / w);
            } else if (wantHeight) {
                height = wantHeight;
                width = w * (wantHeight / h);
            }
            attr = 'width="' + width + '" height="' + height + '"';
        }
    }
    element.innerHTML = '<img src="' + escapeHTML(url) + '" ' + attr + '></img>';
}

function getCover(id, callback) {
    var jsonp = "jsonp" + Math.random().toString().substring(2);
    var head = document.getElementsByTagName("head")[0];
    var script = document.createElement("script");
    script.src = "//ws.gbv.de/covers/?format=seealso&callback="+jsonp+"&id="+id;
    script.type = "text/javascript";
    script.charset = "UTF-8";
    window[jsonp] = function(data) {
        if (data[3] && data[3].length) {
            callback(data[2][0], data[3][0]);
        }
        window[jsonp] = undefined; // GC
        try{ delete window[jsonp]; } catch(e){}
        if (head) script.parentNode.removeChild(script);
    };
    head.appendChild(script);
}

// Ersetze alle Elemente mit class=coverdienst
var elements = document.getElementsByClassName('coverdienst');
for (i=0; i<elements.length; i++) {
  var e = elements[i];
  var id = e.getAttribute("title") || "";
  id = id.replace(/^\s+|\s+$/g,"");
  getCover( id, function(label, size, url) {
    displayCover(e, label, size, url);
  });
}
